book-content/chapters/12-the-weird-parts.md

```diff
@@ -0,0 +1,1653 @@
+
+```ts twoslash
+// @errors: 2345
+class Song {
+  title: string;
+  artist: string;
+
+  constructor(title: string, artist: string) {
+    this.title = title;
+    this.artist = artist;
+  }
+}
+
+const playSong = (song: Song) =>
+  console.log(`Playing ${song.title} by ${song.artist}`);
+
+// ---cut---
+const song1 = new Song("Song 1", "Artist 1");
+
+playSong(song1);
+
+playSong(Song);
+```
+
+In this case, TypeScript shows an error when we try to pass the `Song` class itself to the `playSong` function. This is because `Song` is a class, and not an instance of the class.
+
+So, classes exists in both the type and value worlds, and represents an instance of the class when used as a type.
+
+### Enums
+
+Enums can also cross between worlds.
+
+Consider this `AlbumStatus` enum, and a function that determines whether a discount is available:
+
+```tsx
+enum AlbumStatus {
+  NewRelease = 0,
+  OnSale = 1,
+  StaffPick = 2,
+  Clearance = 3,
+}
+
+function logAlbumStatus(status: AlbumStatus) {
+  if (status === AlbumStatus.NewRelease) {
+    console.log("No discount available.");
+  } else {
+    console.log("Discounted price available.");
+  }
+}
+```
+
+You could use `typeof AlbumStatus` to refer to the entire structure of the enum itself:
+
+```typescript
+function logAlbumStatus(status: typeof AlbumStatus) {
+  // ...implementation
+}
+```
+
+But then you'd need to pass in a structure matching the enum to the function:
+
+```typescript
+logAlbumStatus({
+  NewRelease: 0,
+  OnSale: 1,
+  StaffPick: 2,
+  Clearance: 3,
+});
+```
+
+When used as a type, enums refer to the members of the enum, not the entire enum itself.
+
+### The `this` Keyword
+
+The `this` keyword can also cross between the type and value worlds.
+
+To illustrate, we'll work with this `Song` class that has a slightly different implementation than the one we saw earlier:
+
+```typescript
+class Song {
+  playCount: number;
+
+  constructor(title: string) {
+    this.playCount = 0;
+  }
+
+  play(): this {
+    this.playCount += 1;
+    return this;
+  }
+}
+```
+
+Inside of the `play` method, `this.playCount` uses `this` as a value, to access the `this.playCount` property, but also as a type, to type the return value of the method.
+
+When the `play` method returns `this`, in the type world it signifies that the method returns an instance of the current class.
+
+This means that we can create a new `Song` instance and chain multiple calls to the `play` method:
+
+```tsx
+const earworm = new Song("Mambo No. 5", "Lou Bega").play().play().play();
+```
+
+`this` is a rare case where `this` and `typeof this` are the same thing. We could replace the `this` return type with `typeof this` and the code would still work the same way:
+
+```typescript
+class Song {
+  // ...implementation
+
+  play(): typeof this {
+    this.playCount += 1;
+    return this;
+  }
+}
+```
+
+Both point to the current instance of the class.
```

Comment:
It could be helpful to mention what that means more. 

With inheritance `this` will refer to actual current instance of the class that might be derived from this declaring class:
```ts
class Song {
  play(): typeof this {
    return this;
  }
}

class RapSong extends Song {}

const oneBeer = new RapSong()

oneBeer.play() // RapSong
```